### PR TITLE
feat: Implement AffineToNeura pass with loop nest analysis and valid signal optimization

### PR DESCRIPTION
**Pull request by guosran**
## Overview

This PR implements the `AffineToNeura` conversion pass to lower Affine dialect operations to Neura dialect for CGRA execution.

#### 1. Loop Nest Analysis
Introduces `LoopNestAnalysis` that:
- Builds loop hierarchy trees to track parent-child relationships
- Identifies perfect vs imperfect nesting patterns
- Calculates nesting depth for each loop

#### 2. Valid Signal Optimization
Child loops reuse parent's valid signal instead of creating redundant control signals:
- **Perfect nests**: All nested levels share one `grant_once` at top level
- **Imperfect nests**: Still reuse parent's valid signal where semantically correct
- **Independent loops**: Each gets its own `grant_once` for proper isolation

#### 3. Affine Expression Expansion
Recursively expands complex affine expressions into explicit Neura operations:
- Arithmetic: `Add`, `Mul`, `Sub`, `Div`, `Rem`
- Supports nested expressions: `(d0 + d1) * 2` → explicit operation chain
- Handles `CeilDiv` via formula: `ceildiv(a,b) = floordiv(a+b-1, b)`

#### 4. Pattern-Based Conversion
Uses MLIR's Dialect Conversion framework with patterns for:
- `affine.load` → `neura.load_indexed`
- `affine.store` → `neura.store_indexed`
- `affine.apply` → Neura arithmetic ops
- `affine.for` → `neura.loop_control` with optimized valid signals

**Test Coverage**:
- 6 focused test suites covering various loop patterns
- Tests include: perfect nesting, imperfect nesting, corner cases, complex expressions, deep nesting, and branch operations (currently not supported, shows how to lower with a premature alternative approach)

### CODE REVIEW

File: lib/NeuraDialect/Mapping/mapping_util.cpp [hunk 1]
```diff
@@ -961,8 +982,14 @@ bool mlir::neura::placeAndRoute(Operation *op, const MappingLoc &target_loc,
         continue;
       }
       Operation *data_move = operand.getDefiningOp();
-      assert(isa<neura::DataMovOp>(data_move) &&
-             "Expected a DataMovOp as operand producer");
+      
+      // In steering mode, some operands may not be DataMovOp (e.g., constants, carry, etc.)
+      if (!isa<neura::DataMovOp>(data_move)) {
+        // Skip non-DataMovOp operands in steering mode
```
Comment (tancheng) [inline lib/NeuraDialect/Mapping/mapping_util.cpp]: `// Skip` -> `// Skips`. and end with period.

File: lib/NeuraDialect/Transforms/MapToAcceleratorPass.cpp [hunk 2]
```diff
@@ -512,18 +806,27 @@ struct MapToAcceleratorPass
       bool is_steering_mode =
           (dataflow_mode_attr && dataflow_mode_attr.getValue() == "steering");
 
-      // If steering mode, enforce spatial-only mapping.
+      // If steering mode, enforce spatial-only mapping unless explicitly allowed.
```
Comment (tancheng) [inline lib/NeuraDialect/Transforms/MapToAcceleratorPass.cpp]: What about eliminate "--allow-steering-spatial-temporal", why not just use "spatial-temporal"/"spatial-only" to indicate whether steering is on which mapping strategy?

File: lib/Conversion/AffineToNeura/AffineToNeuraPass.cpp [hunk 3]
```diff
@@ -0,0 +1,388 @@
...
+
+    while (parent_op) {
+      if (isa<affine::AffineForOp>(parent_op)) {
+        has_affine_for_parent = true;
+        break;
+      }
+      parent_op = parent_op->getParentOp();
+    }
+
+    // If it has no AffineForOp parent, it's a Ftop-level loop
```
Comment (tancheng) [inline lib/Conversion/AffineToNeura/AffineToNeuraPass.cpp]: what is "Ftop"?

File: lib/Conversion/AffineToNeura/AffineToNeuraPass.cpp [hunk 4]
```diff
@@ -0,0 +1,388 @@
...
+    int64_t lower_bound_constant = for_op.getConstantLowerBound();
+    lower_bound_val = builder.create<neura::ConstantOp>(
+        loc, index_type, builder.getIndexAttr(lower_bound_constant), nullptr); // nullptr is for predicated bit
+  } else {
+    // If the lower bound is not constant, we need to use affine.apply
+    affine::AffineBound lower_bound = for_op.getLowerBound();
+    AffineMap lower_bound_map = lower_bound.getMap();
+    ValueRange lower_bound_operands = for_op.getLowerBoundOperands();
+    lower_bound_val = builder.create<affine::AffineApplyOp>(
+        loc, lower_bound_map, lower_bound_operands);
```
Comment (tancheng) [inline lib/Conversion/AffineToNeura/AffineToNeuraPass.cpp]: Even non-const lower/upper bound, we are able to pattern-match to `loop_control`?

File: lib/Conversion/AffineToNeura/AffineToNeuraPass.cpp [hunk 5]
```diff
@@ -0,0 +1,388 @@
...
+                OpBuilder builder(outer_for_op);
+                if (failed(lowerAffineFor(outer_for_op, builder, mapping))) {
+                  outer_for_op.emitError("[affine2neura] Failed to lower "
+                                         "outermost affine.for operation");
+                  signalPassFailure();
+                }
+              }
+
+              RewritePatternSet patterns(context);
+              patterns.add<AffineLoadLowering, AffineStoreLowering>(context);
```
Comment (tancheng) [inline lib/Conversion/AffineToNeura/AffineToNeuraPass.cpp]: Can these be added outside the walk block?

File: lib/NeuraDialect/Transforms/MapToAcceleratorPass.cpp (hunk 2 above)
Comment (tancheng) [reply lib/NeuraDialect/Transforms/MapToAcceleratorPass.cpp]: Delete this flag but leveraging spatial-temporal/spatial-only as i mentioned? Or disallow the spatial-temporal for steering? Or delete this PR?

Comment (ShangkunLi) [reply lib/NeuraDialect/Transforms/MapToAcceleratorPass.cpp]: Not yet. Still spatial-only for steering.

**Comment by tancheng** [comment]
Is this ready for review?

**Comment by guosran** [comment]
yes

File: lib/NeuraDialect/Mapping/mapping_util.cpp (hunk 1 above)
Comment (guosran) [reply lib/NeuraDialect/Mapping/mapping_util.cpp]: Comment style fixed.

File: lib/Conversion/AffineToNeura/AffineToNeuraPass.cpp (hunk 3 above)
Comment (guosran) [reply lib/Conversion/AffineToNeura/AffineToNeuraPass.cpp]: Sorry, this is a typo. It has been fixed to "top".

File: lib/Conversion/AffineToNeura/AffineToNeuraPass.cpp [hunk 6]
```diff
@@ -0,0 +1,388 @@
...
+
+  builder.setInsertionPointToStart(continue_block);
+
+  for_op.erase();
+
+  return success();
+}
+
+affine::AffineForOp findOuterMostAffineFor(func::FuncOp &func_op) {
+  // Find the outermost affine.for operation
```
[the remark this answers is not in the document]
Comment (guosran) [reply lib/Conversion/AffineToNeura/AffineToNeuraPass.cpp]: Comment style fixed.

File: lib/NeuraDialect/Mapping/mapping_util.cpp [hunk 7]
```diff
@@ -961,8 +982,14 @@ bool mlir::neura::placeAndRoute(Operation *op, const MappingLoc &target_loc,
         continue;
       }
       Operation *data_move = operand.getDefiningOp();
-      assert(isa<neura::DataMovOp>(data_move) &&
-             "Expected a DataMovOp as operand producer");
+      
+      // In steering mode, some operands may not be DataMovOp (e.g., constants, carry, etc.)
+      if (!isa<neura::DataMovOp>(data_move)) {
```
[the remark this answers is not in the document]
Comment (guosran) [reply lib/NeuraDialect/Mapping/mapping_util.cpp]: I've updated the code to use a helper function `is_steering_unwrapped_op()` that explicitly checks for steering-specific operations (ConstantOp, CarryOp, InvariantOp, CarryInvariantOp, ConditionalSelectOp, InvariantGroupOp, ReserveOp).

File: lib/Conversion/AffineToNeura/AffineToNeuraPass.cpp (hunk 4 above)
Comment (guosran) [reply lib/Conversion/AffineToNeura/AffineToNeuraPass.cpp]: Currently we only support constant bounds.

File: lib/Conversion/AffineToNeura/AffineToNeuraPass.cpp (hunk 5 above)
Comment (guosran) [reply lib/Conversion/AffineToNeura/AffineToNeuraPass.cpp]: Fixed. Pattern registration moved outside the walk block for better performance and cleaner code structure. Patterns are now registered once per function instead of repeatedly during traversal.

File: lib/Conversion/AffineToNeura/AffineToNeuraPass.cpp [hunk 8]
```diff
@@ -0,0 +1,388 @@
...
+    MLIRContext *context = module_op.getContext();
+    IRMapping mapping;
+    module_op.walk(
+        [&](func::FuncOp func_op) {
+          if (func_op->hasAttr(mlir::accel::kAcceleratorAttr)) {
+            auto target = func_op->getAttrOfType<StringAttr>(
+                mlir::accel::kAcceleratorAttr);
+            if (target && target.getValue() == mlir::accel::kNeuraTarget) {
+              while (affine::AffineForOp outer_for_op =
+                         findOuterMostAffineFor(func_op)) {
```
[the remark this answers is not in the document]
Comment (guosran) [reply lib/Conversion/AffineToNeura/AffineToNeuraPass.cpp]: I have uploaded the code here in this PR. Sorry for the confusion.

**Comment by guosran** [comment]
> > yes
> 
> Can you reply on each of my previous comments so I would know what is happening since then?


Replied. You may also check the updated overview for a generalized description.

File: include/Conversion/AffineToNeura/LoopNestAnalysis.h [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+namespace mlir {
+namespace neura {
+
+/// Loop information structure - Stores all analysis information for a single loop.
+struct LoopInfo {
+  affine::AffineForOp loop;              // The loop operation itself.
+  LoopInfo *parent = nullptr;            // Parent loop (nullptr if top-level).
+  llvm::SmallVector<LoopInfo *, 4> children;  // Child loops list.
+  unsigned depth = 0;                    // Nesting depth (0=top-level).
+  bool isPerfectNest = true;             // Whether it is a perfect nest.
```
Comment (tancheng) [inline include/Conversion/AffineToNeura/LoopNestAnalysis.h]: Let's use snake_case for all variables, i.e., `isPerfectNest` -> `is_perfect_nest`.

File: include/NeuraDialect/NeuraOps.td [hunk 10]
```diff
@@ -689,4 +657,133 @@ def Neura_InvariantOp : Op<NeuraDialect, "invariant">{
...
+    single atomic operation to reduce RecMII.
+    
+    Semantics:
+    - Performs comparison: result = (lhs <predicate> rhs)
+    - If result is false: return value
+    - If result is true: return default value (typically from hardware)
+    
+    Replaces the pattern:
+      %cond = neura.icmp %lhs, %rhs <{cmpType = "slt"}>
+      %result = neura.false_steer %value, %cond
```
Comment (tancheng) [inline include/NeuraDialect/NeuraOps.td]: Will there be a `true_steer`? can we also show how these two are generated from a high level IR in this example?

File: lib/Conversion/AffineToNeura/AffineToNeuraPass.cpp [hunk 11]
```diff
@@ -0,0 +1,477 @@
...
+
+using namespace mlir;
+using namespace mlir::neura;
+using namespace mlir::func;
+
+#define GEN_PASS_DEF_LOWERAFFINETONEURA
+#include "Conversion/ConversionPasses.h.inc"
+
+namespace {
+LogicalResult convertAffineMapToIndices(AffineMap map, ValueRange map_operands,
```
Comment (tancheng) [inline lib/Conversion/AffineToNeura/AffineToNeuraPass.cpp]: Thanks @guosran for the effort. Can you provide example as comment for each conversion in this file, so we can understand what would happen?

File: test/Conversion/AffineToNeura/complex-affine-expressions.mlir [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+  // CHECK-NEXT: %[[I:.*]], %[[VALID:.*]] = "neura.loop_control"(%[[GRANT]]) <{end = 10 : i64, iterationType = "increment", start = 0 : i64, step = 1 : i64}> : (i1) -> (index, i1)
+  // CHECK-NEXT: %[[C2:.*]] = "neura.constant"() <{value = 2 : index}> : () -> index
+  // CHECK-NEXT: %[[MUL:.*]] = "neura.mul"(%[[I]], %[[C2]]) : (index, index) -> index
+  // CHECK-NEXT: %[[C1:.*]] = "neura.constant"() <{value = 1 : index}> : () -> index
+  // CHECK-NEXT: %[[ADD:.*]] = "neura.add"(%[[MUL]], %[[C1]]) : (index, index) -> index
+  // CHECK-NEXT: %{{.*}} = neura.load_indexed %arg0[%[[ADD]] : index] memref<100xf32> : f32
+  // CHECK-NEXT: return
+  func.func @complex_expression(%arg0: memref<100xf32>) {
+    affine.for %i = 0 to 10 {
+      %0 = affine.load %arg0[2 * %i + 1] : memref<100xf32>
```
Comment (tancheng) [inline test/Conversion/AffineToNeura/complex-affine-expressions.mlir]: Can you point me to the line how the calculation on index is captured/analyzed?

File: include/Conversion/AffineToNeura/LoopNestAnalysis.h (hunk 9 above)
Comment (guosran) [reply include/Conversion/AffineToNeura/LoopNestAnalysis.h]: fixed

File: include/NeuraDialect/NeuraOps.td [hunk 13]
```diff
@@ -689,4 +657,133 @@ def Neura_InvariantOp : Op<NeuraDialect, "invariant">{
...
+    
+    Semantics:
+    - If inner_condition is false (first inner iteration): return initial value
+    - Else if outer_condition is false (outer loop active, inner loop invariant): 
+        return initial value from inner carry
+    - Else: return carried value
+    
+    Replaces the pattern:
+      %carry_result = neura.carry %init, %inner_cond, %carried
+      %inv_result = neura.invariant %carry_result, %outer_cond
```
[the remark this answers is not in the document]
Comment (guosran) [reply include/NeuraDialect/NeuraOps.td]: Sorry these are wrong and redundant operations. I have deleted them.

File: include/NeuraDialect/NeuraOps.td (hunk 10 above)
Comment (guosran) [reply include/NeuraDialect/NeuraOps.td]: Sorry these are wrong and redundant operations. I have deleted them.

File: lib/NeuraDialect/Mapping/mapping_util.cpp [hunk 14]
```diff
@@ -87,6 +92,14 @@ bool is_non_materialized(Operation *op) {
   return mlir::isa<neura::ReserveOp, neura::CtrlMovOp, neura::DataMovOp>(op);
 }
 
+// Returns true if the operation is a steering-mode operation that doesn't
+// require DataMovOp wrapping (e.g., constants, carry, invariant, etc.).
+bool is_steering_unwrapped_op(Operation *op) {
+  return mlir::isa<neura::ConstantOp, neura::CarryOp, neura::InvariantOp,
```
[the remark this answers is not in the document]
Comment (guosran) [reply lib/NeuraDialect/Mapping/mapping_util.cpp]: Because constants don't have data dependencies that require routing.

File: test/Conversion/AffineToNeura/complex-affine-expressions.mlir [hunk 15]
```diff
@@ -0,0 +1,90 @@
+// RUN: mlir-neura-opt %s --lower-affine-to-neura | FileCheck %s
+
+// This test verifies that complex affine expressions are correctly expanded
+// into explicit Neura arithmetic operations.
+
+module {
+  // Test 1: Multiplication expression (d0 * 2)
+  // CHECK-LABEL: func.func @mul_expression
+  // CHECK-NEXT: %[[GRANT:.*]] = "neura.grant_once"() : () -> i1
+  // CHECK-NEXT: %[[I:.*]], %[[VALID:.*]] = "neura.loop_control"(%[[GRANT]]) <{end = 10 : i64, iterationType = "increment", start = 0 : i64, step = 1 : i64}> : (i1) -> (index, i1)
```
[the remark this answers is not in the document]
Comment (guosran) [reply test/Conversion/AffineToNeura/complex-affine-expressions.mlir]: The `index` type is MLIR's built-in type for array indexing and loop induction variables, which automatically maps to the target's pointer size (32-bit or 64-bit). This matches the Affine dialect convention where `affine.for` induction variables are typed as `index`.

File: test/Conversion/AffineToNeura/complex-affine-expressions.mlir [hunk 16]
```diff
@@ -0,0 +1,90 @@
...
+  // CHECK-NEXT: %[[C2:.*]] = "neura.constant"() <{value = 2 : index}> : () -> index
+  // CHECK-NEXT: %[[MUL:.*]] = "neura.mul"(%[[I]], %[[C2]]) : (index, index) -> index
+  // CHECK-NEXT: %{{.*}} = neura.load_indexed %arg0[%[[MUL]] : index] memref<10xf32> : f32
```
[the remark this answers is not in the document]
Comment (guosran) [reply test/Conversion/AffineToNeura/complex-affine-expressions.mlir]: `neura.load` takes a pre-computed pointer address (from LLVM dialect), while `neura.load_indexed` takes base + indices for multi-dimensional array access (from Affine/MemRef dialects) and integrates address calculation internally. They target different lowering paths, and keeping them separate helps with pattern matching and optimization opportunities in later passes.

File: test/Conversion/AffineToNeura/complex-affine-expressions.mlir (hunk 12 above)
Comment (guosran) [reply test/Conversion/AffineToNeura/complex-affine-expressions.mlir]: The index calculation is captured and analyzed in the `convertAffineMapToIndices()` function (in `lib/Conversion/AffineToNeura/AffineToNeuraPass.cpp`). It recursively expands AffineExpr into explicit Neura arithmetic operations via the `expandExpr` lambda, which pattern-matches on expression types (constant/dim/binary ops) and generates corresponding neura.constant/add/mul/rem/div operations. For example, `2*%i+1` becomes: constant(2) → mul(%i, 2) → constant(1)

File: test/Conversion/AffineToNeura/complex-affine-expressions.mlir [hunk 17]
```diff
@@ -0,0 +1,90 @@
...
+  // CHECK-NEXT: %[[DIV:.*]] = "neura.div"(%[[I]], %[[C4_1]]) : (index, index) -> index
+  // CHECK-NEXT: %[[C4_2:.*]] = "neura.constant"() <{value = 4 : index}> : () -> index
+  // CHECK-NEXT: %[[REM:.*]] = "neura.rem"(%[[I]], %[[C4_2]]) : (index, index) -> index
+  // CHECK-NEXT: %{{.*}} = neura.load_indexed %arg0[%[[DIV]], %[[REM]] : index, index] memref<8x8xf32> : f32
+  // CHECK-NEXT: return
+  func.func @floordiv_expression(%arg0: memref<8x8xf32>) {
+    affine.for %i = 0 to 32 {
+      %row = affine.apply affine_map<(d0) -> (d0 floordiv 4)>(%i)
+      %col = affine.apply affine_map<(d0) -> (d0 mod 4)>(%i)
+      %0 = affine.load %arg0[%row, %col] : memref<8x8xf32>
```
[the remark this answers is not in the document]
Comment (guosran) [reply test/Conversion/AffineToNeura/complex-affine-expressions.mlir]: Yes. `memref<8x8xf32>` is the input type (base pointer type), and the output type is f32. The syntax `affine.load %arg0[%row, %col] : memref<8x8xf32>` implicitly returns f32.

The `affine.apply` operations are expanded into explicit Neura arithmetic operations by the AffineApplyLowering pattern (in AffineToNeuraPass.cpp):

```
affine.apply affine_map<(d0) -> (d0 floordiv 4)>(%i) → %div = neura.div %i, %c4
affine.apply affine_map<(d0) -> (d0 mod 4)>(%i) → %rem = neura.rem %i, %c4
```
The lowering recursively expands AffineExpr into constant creation + arithmetic ops.

File: test/Conversion/AffineToNeura/complex-affine-expressions.mlir [hunk 18]
```diff
@@ -0,0 +1,90 @@
...
+  }
+
+  // Test 5: Multiple dimensions with complex expressions
+  // CHECK-LABEL: func.func @multi_dim_complex
+  // CHECK-NEXT: %[[GRANT:.*]] = "neura.grant_once"() : () -> i1
+  // CHECK-NEXT: %[[I:.*]], %[[VALID_I:.*]] = "neura.loop_control"(%[[GRANT]]) <{end = 10 : i64, iterationType = "increment", start = 0 : i64, step = 1 : i64}> : (i1) -> (index, i1)
+  // CHECK-NEXT: %[[J:.*]], %[[VALID_J:.*]] = "neura.loop_control"(%[[VALID_I]]) <{end = 20 : i64, iterationType = "increment", start = 0 : i64, step = 1 : i64}> : (i1) -> (index, i1)
+  // CHECK-NEXT: %[[C1:.*]] = "neura.constant"() <{value = 1 : index}> : () -> index
+  // CHECK-NEXT: %[[ADD:.*]] = "neura.add"(%[[J]], %[[C1]]) : (index, index) -> index
+  // CHECK-NEXT: %{{.*}} = neura.load_indexed %arg0[%[[I]], %[[ADD]] : index, index] memref<10x20xf32> : f32
```
[the remark this answers is not in the document]
Comment (guosran) [reply test/Conversion/AffineToNeura/complex-affine-expressions.mlir]: No explicit dimension limit is enforced here. The lowering supports arbitrary dimensions.

File: test/Conversion/AffineToNeura/loop-nest-optimization.mlir [hunk 19]
```diff
@@ -0,0 +1,98 @@
...
+    affine.for %j = 0 to 20 {
+      %v = affine.load %A[%i, %j] : memref<10x20xf32>
+    }
+  }
+  return
+}
+
+// Test 4: Two separate top-level loops - each should get its own grant_once
+// CHECK-LABEL: func.func @two_top_level_loops
+// CHECK-NEXT: %[[GRANT1:.*]] = "neura.grant_once"() : () -> i1
```
[the remark this answers is not in the document]
Comment (guosran) [reply test/Conversion/AffineToNeura/loop-nest-optimization.mlir]: Both parameters of `rant_once` (`value` and `constant_value`) are optional. In loop control scenarios, we don't pass in any operand, only generating a pure `i1` control signal. In dataflow scenarios, however, we can pass in a value to indicate its validity.

File: lib/Conversion/AffineToNeura/AffineToNeuraPass.cpp [hunk 20]
```diff
@@ -0,0 +1,388 @@
...
+
+struct AffineApplyLowering : public OpRewritePattern<affine::AffineApplyOp> {
+  using OpRewritePattern<affine::AffineApplyOp>::OpRewritePattern;
+  LogicalResult matchAndRewrite(affine::AffineApplyOp apply_op,
+                                PatternRewriter &rewriter) const override {
+    AffineMap map = apply_op.getAffineMap();
+    ValueRange operands = apply_op.getMapOperands();
+    Location loc = apply_op.getLoc();
+
+    if (map.getNumResults() != 1) {
```
[the remark this answers is not in the document]
Comment (guosran) [reply lib/Conversion/AffineToNeura/AffineToNeuraPass.cpp]: Added, sorry for the confusion.

File: test/Conversion/AffineToNeura/single-iteration.mlir [hunk 21]
```diff
@@ -0,0 +1,23 @@
...
+// Expected output after --lower-affine-to-neura transformation:
+// Verify: 1) no affine ops, 2) all neura ops present, 3) exact IR match
+// ============================================================================
+// CHECK-LABEL: func.func @single_iteration(%arg0: memref<1xf32>)
+// CHECK-NEXT: %[[V0:.*]] = "neura.grant_once"() : () -> i1
+// CHECK-NEXT: %[[NEXT:.*]], %[[VALID:.*]] = "neura.loop_control"(%[[V0]]) <{end = 1 : i64, iterationType = "increment", start = 0 : i64, step = 1 : i64}> : (i1) -> (index, i1)
+// CHECK-NEXT: %{{.*}} = neura.load_indexed %arg0[%[[NEXT]] : index] memref<1xf32> : f32
+// CHECK-NEXT: return
+// CHECK-NEXT: }
+// CHECK-NOT: affine.
```
[the remark this answers is not in the document]
Comment (guosran) [reply test/Conversion/AffineToNeura/single-iteration.mlir]: Isuue created.

File: test/Conversion/AffineToNeura/complex-affine-expressions.mlir (hunk 12 above)
Comment (tancheng) [reply test/Conversion/AffineToNeura/complex-affine-expressions.mlir]: Excellent~!

File: test/Conversion/AffineToNeura/complex-affine-expressions.mlir (hunk 18 above)
[the remark this answers is not in the document]
Comment (tancheng) [reply test/Conversion/AffineToNeura/complex-affine-expressions.mlir]: Yeah, but I am thinking from HW. We might not able to support 100 operands. There should be a limit. Maybe we can add constraint later once the initial HW is implemented.

File: test/Conversion/AffineToNeura/imperfect-ops-after.mlir [hunk 22]
```diff
@@ -0,0 +1,29 @@
...
+
+// Imperfect Nesting: Operations after child loop
+module {
+  func.func @imperfect_ops_after(%arg0: memref<10x20xf32>, %arg1: memref<10xf32>) {
+    affine.for %i = 0 to 10 {
+      affine.for %j = 0 to 20 {
+        %0 = affine.load %arg0[%i, %j] : memref<10x20xf32>
+      }
+      %cst = arith.constant 1.0 : f32
+      affine.store %cst, %arg1[%i] : memref<10xf32>
```
[the remark this answers is not in the document]
Comment (tancheng) [reply test/Conversion/AffineToNeura/imperfect-ops-after.mlir]: Sorry my bad. We basically flatten them, so there shouldn't be indent. Maybe you can use empty line (`//`) to indicate there two regions.

File: test/Conversion/AffineToNeura/unsupported-affine-if.mlir [hunk 23]
```diff
@@ -0,0 +1,29 @@
...
+// 2. Error is emitted indicating failed legalization
+// 3. Affine operations remain unchanged in the IR
+//
+// CHECK: error:
+// CHECK: affine.if
+//
+// Note: affine.if is not currently supported in this lowering pass.
+// Conditional execution would require predicated operations or 
+// control flow handling in the dataflow model.
+// ============================================================================
```
Comment (tancheng) [inline test/Conversion/AffineToNeura/unsupported-affine-if.mlir]: Can we then also use `lower-affine pass` to transform these into `scf`, then lower to `llvm.br` etc, then map to neura in this example?

File: test/Conversion/AffineToNeura/loop-nest-optimization.mlir (hunk 19 above)
[the remark this answers is not in the document]
Comment (guosran) [reply test/Conversion/AffineToNeura/loop-nest-optimization.mlir]: Fixed. Thanks for pointing it out.

File: test/Conversion/AffineToNeura/unsupported-affine-if.mlir (hunk 23 above)
Comment (guosran) [reply test/Conversion/AffineToNeura/unsupported-affine-if.mlir]: Edited.

File: lib/NeuraDialect/Mapping/mapping_util.cpp [hunk 24]
```diff
@@ -87,6 +87,14 @@ bool is_non_materialized(Operation *op) {
   return mlir::isa<neura::ReserveOp, neura::CtrlMovOp, neura::DataMovOp>(op);
 }
 
+// Returns true if the operation is a steering-mode operation that doesn't
+// require DataMovOp wrapping (e.g., carry, invariant, reserve).
```
Comment (tancheng) [inline lib/NeuraDialect/Mapping/mapping_util.cpp]: I am curious why `carry` and `invariant` do not need `data_move` for mapping @ShangkunLi

File: test/Conversion/AffineToNeura/unsupported-affine-if.mlir (hunk 23 above)
Comment (tancheng) [reply test/Conversion/AffineToNeura/unsupported-affine-if.mlir]: Can we call these passes using `mlir-neura-opt --xxx` and `// CHECK` the output IRs here to demonstrate what would happen?

File: test/Conversion/AffineToNeura/unsupported-dynamic-bounds.mlir [hunk 25]
```diff
@@ -0,0 +1,27 @@
+// RUN: not mlir-neura-opt %s --lower-affine-to-neura 2>&1 | FileCheck %s
```
Comment (tancheng) [inline test/Conversion/AffineToNeura/unsupported-dynamic-bounds.mlir]: What does `not` mean here?

File: test/Conversion/AffineToNeura/loop-nest-optimization.mlir [hunk 26]
```diff
@@ -0,0 +1,98 @@
+// RUN: mlir-neura-opt %s --lower-affine-to-neura | FileCheck %s
+
+// Test 1: Perfect nested loops - should reuse valid signals
+// CHECK-LABEL: func.func @perfect_nest_2d
+// CHECK-NEXT: %[[TRUE:.*]] = "neura.constant"() <{value = true}> : () -> i1
+// CHECK-NEXT: %[[I:.*]], %[[VALID_I:.*]] = "neura.loop_control"(%[[TRUE]]) <{end = 10 : i64, iterationType = "increment", start = 0 : i64, step = 1 : i64}> : (i1) -> (index, i1)
+// CHECK-NEXT: %[[J:.*]], %[[VALID_J:.*]] = "neura.loop_control"(%[[VALID_I]]) <{end = 20 : i64, iterationType = "increment", start = 0 : i64, step = 1 : i64}> : (i1) -> (index, i1)
+// CHECK-NEXT: %{{.*}} = neura.load_indexed %arg0[%[[I]], %[[J]] : index, index] memref<10x20xf32> : f32
+// CHECK-NEXT: return
```
[the remark this answers is not in the document]
Comment (guosran) [reply test/Conversion/AffineToNeura/loop-nest-optimization.mlir]: fixed

File: test/Conversion/AffineToNeura/unsupported-dynamic-bounds.mlir (hunk 25 above)
Comment (guosran) [reply test/Conversion/AffineToNeura/unsupported-dynamic-bounds.mlir]: Deleted this file, since we do not plan to support dynamic bound at current stage.

File: lib/NeuraDialect/Mapping/mapping_util.cpp [hunk 27]
```diff
@@ -87,6 +87,14 @@ bool is_non_materialized(Operation *op) {
...
+// Note: ConstantOp is NOT included here because constants DO need routing
```
[the remark this answers is not in the document]
Comment (guosran) [reply lib/NeuraDialect/Mapping/mapping_util.cpp]: It's "do" rather than "d0". I have deleted this line to avoid confusion.

File: lib/NeuraDialect/Mapping/mapping_util.cpp (hunk 24 above)
Comment (guosran) [reply lib/NeuraDialect/Mapping/mapping_util.cpp]: Deleted to align `InsertDataMovPass` behavior

File: lib/NeuraDialect/Mapping/mapping_util.cpp [hunk 28]
```diff
@@ -87,6 +87,12 @@ bool is_non_materialized(Operation *op) {
   return mlir::isa<neura::ReserveOp, neura::CtrlMovOp, neura::DataMovOp>(op);
 }
 
+// Returns true if the operation doesn't require DataMovOp wrapping.
+// This must match InsertDataMovPass behavior which only skips ReserveOp.
+bool is_steering_unwrapped_op(Operation *op) {
+  return mlir::isa<neura::ReserveOp>(op);
+}
+
```
Comment (tancheng) [inline lib/NeuraDialect/Mapping/mapping_util.cpp]: Then we can reuse above `is_non_materialized()`?

File: test/Conversion/AffineToNeura/loop-nest-optimization.mlir [hunk 29]
```diff
@@ -0,0 +1,98 @@
...
+func.func @perfect_nest_2d(%A: memref<10x20xf32>) {
+  affine.for %i = 0 to 10 {
+    affine.for %j = 0 to 20 {
+      %v = affine.load %A[%i, %j] : memref<10x20xf32>
+    }
+  }
+  return
+}
+// CHECK-LABEL: func.func @perfect_nest_2d
+// CHECK-NEXT: %[[TRUE:.*]] = "neura.constant"() <{value = true}> : () -> i1
```
Comment (ShangkunLi) [inline test/Conversion/AffineToNeura/loop-nest-optimization.mlir]: Could you please avoid using non-deterministic symbols in the check files?

Comment (guosran) [reply test/Conversion/AffineToNeura/loop-nest-optimization.mlir]: fixed

Comment (tancheng) [reply test/Conversion/AffineToNeura/loop-nest-optimization.mlir]: Shouldn't we resolve this and merge this PR?
